### DISCUSSION TITLE
Combining workflow triggers (specifically schedule + pull_request)

### DISCUSSION THREAD

**Discussion by BenHenning** [community/community#55768] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi,

I'm trying to figure out if I can set up one of my project's actions workflows to run no more than once per day, and only after at least one push to an active PR. The workflow is optional, but quite expensive to run so I'd rather not have it consuming actions resources for every push to a branch; running once per day is optimal. In order for this to work, the cron should be scheduled such that it can't run more than once even if multiple pushes occur.

From looking at the documentation, it seems that ``schedule`` and ``pull_request`` trigger independently. Because of that, I think I'd need to use something like workflow_dispatch in conjunction with a scheduled job (e.g. something like this: https://cardinalby.github.io/blog/post/github-actions/implementing-deferred-steps/) but the thing I'm not sure about is: my job requires accessing the PR information (since it's performing a comparison of the latest PR commit with its base commit, and then posting a comment on the PR thread). Will that even work in a deferred workflow?

Hoping to get some suggestions on how I can even approach this.

Thanks!

**Comment by airtower-luna** [comment, marked as answer]
> my job requires accessing the PR information (since it's performing a comparison of the latest PR commit with its base commit, and then posting a comment on the PR thread). Will that even work in a deferred workflow?

It's definitely possible, depending on the [`permissions` setting](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions) you give the workflow. What you won't have in a scheduled workflow is PR data in `github.event`, instead you'll have to retrieve the PR information from the API (authorized with the `GITHUB_TOKEN`, that's why the permissions are important).

**Reply by BenHenning** [reply]
Thanks--I think I understand how the permissions part is necessary (and possibly how to do it--I'll probably need to play with an actual setup to get a feel for it).

Might you have any ideas on how I can solve the "only run once per day" part? I'm concerned that simply scheduling a cron for my job on each PR push will result in there being a new workflow to run each day, rather than uniquifying and only running one. Perhaps I can use something like ``gh`` to ensure there's not already a cron workflow of that type running? I'm not sure on the specifics of how to do that, but it seems possible.

**Reply by airtower-luna** [reply]
> I'm concerned that simply scheduling a cron for my job on each PR push will result in there being a new workflow to run each day

Yes, that sounds like it'd be an extremely bad idea. :sweat_smile: 

Based on what you've described so far I'd just schedule one workflow to run every day, fixed (not from a PR or anything). And then let that workflow check if there are PRs that need doing whatever it is, and iterate over them to do it.

One thing to keep in mind: Scheduled workflows [aren't guaranteed to run at an exact time](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule), you might want to consider that when writing the code to detect which PRs need to be handled.

**Reply by BenHenning** [reply]
Ah, that actually sounds like a less brittle approach (to have the crons run independently). I like that idea, and will dig into that. Thanks much for the suggestions!
